src/tests/unit/test_icmp.py

```diff
@@ -312,3 +318,110 @@ def _iptables_parse_supported_icmp_types_6():
 def test_iptables_parse_supported_icmp_types():
     _iptables_parse_supported_icmp_types_4()
     _iptables_parse_supported_icmp_types_6()
+
+
+###############################################################################
+
+
+def _get_destination(xmlobj):
+    d = xmlobj.destination
+    if d == []:
+        d = ["ipv4", "ipv6"]
+
+    assert d in (
+        ["ipv4"],
+        ["ipv6"],
+        ["ipv4", "ipv6"],
+    )
+
+    return d
+
+
+def _icmptypes_load_file(dirname, file):
+    assert dirname
+    assert file
+    assert file.endswith(".xml")
+    assert "/" not in file
+
+    full_name = os.path.join(dirname, file)
+
+    assert os.path.exists(dirname)
+    assert os.path.exists(full_name)
+
+    xmlobj = firewall.core.io.icmptype.icmptype_reader(file, dirname)
+    assert xmlobj
+
+    assert xmlobj.name == file[: -len(".xml")]
+    assert xmlobj.path == dirname
+    _get_destination(xmlobj)
+    return xmlobj
+
+
+def test_icmptypes():
+    dirname = helpers.srcdir("config/icmptypes")
+    files = [f for f in os.listdir(dirname) if f.endswith(".xml")]
+    assert files
+    xmlobjs = []
+    for file in files:
+        xmlobjs.append(_icmptypes_load_file(dirname, file))
+
+    nft = firewall.core.nftables.ICMP_TYPES_FRAGMENTS
+    assert set(nft.keys()) == set(["ipv4", "ipv6"])
+
+    # Check that all .xml files are listed in firewall.core.nftables.ICMP_TYPES_FRAGMENTS
+    for xmlobj in xmlobjs:
+        for ipx in _get_destination(xmlobj):
+            assert xmlobj.name in nft[ipx], (
+                f'XML file "{xmlobj.path}/{xmlobj.name}.xml" has no entry '
+                f'"{xmlobj.name}" in firewall.core.nftables.ICMP_TYPES_FRAGMENTS["{ipx}"]'
+            )
+
+    # Check that all firewall.core.nftables.ICMP_TYPES_FRAGMENTS have an .xml file.
+    for ipx in nft:
+        assert ipx in ["ipv4", "ipv6"]
+        for icmp_type in nft[ipx]:
+            l = [xmlobj for xmlobj in xmlobjs if xmlobj.name == icmp_type]
+            assert len(l) == 1
+            xmlobj = l[0]
+            assert ipx in _get_destination(xmlobj)
+
+    # firewall.core.icmp is only used by src/firewall/core/io/ipset.py, and
+    # hard-codes valid ICMP types. It's not used anywhere else.
+    #
+    # It should still correspond to our XML files, where every entry here has
+    # an XML file, but some XML files are not found here.
```

Comment:
I didn't even know we supported ICMP in the `port` field of an ipset. 🤷

Reply:
https://github.com/firewalld/firewalld/pull/1172 touches that area too (commit 2d1f4b3d2c9842cde5b8fb6c25c4e4b7363311d1)